src/Transport/Sending/MessageDispatcher.cs

```diff
@@ -158,28 +157,31 @@ async Task DispatchBatchForDestination(string destination, ServiceBusClient? cli
                     logBuilder = new StringBuilder();
                 }
 
-                var peekedMessage = messagesToSend.Peek();
-                if (messageBatch.TryAddMessage(peekedMessage))
+                var dequeueMessage = messagesToSend.Dequeue();
+                if (messageBatch.TryAddMessage(dequeueMessage))
                 {
-                    var added = messagesToSend.Dequeue();
                     if (Log.IsDebugEnabled)
                     {
-                        added.ApplicationProperties.TryGetValue(Headers.MessageId, out var messageId);
-                        logBuilder!.Append($"{messageId ?? added.MessageId},");
+                        dequeueMessage.ApplicationProperties.TryGetValue(Headers.MessageId, out var messageId);
+                        logBuilder!.Append($"{messageId ?? dequeueMessage.MessageId},");
                     }
                 }
                 else
                 {
-                    peekedMessage.ApplicationProperties.TryGetValue(Headers.MessageId, out var messageId);
-                    var message =
-                        @$"Unable to add the message '#{messageCount - messagesToSend.Count}' with message id '{messageId ?? peekedMessage.MessageId}' to the the batch '#{batchCount}'.
-The message may be too large, or the batch size has reached the maximum allowed messages per batch for the current tier selected for the namespace '{sender.FullyQualifiedNamespace}'.
-To mitigate this problem, either reduce the message size by using the data bus, upgrade to a higher Service Bus tier, or increase the maximum message size.
-If the maximum message size is increased, the endpoint must be restarted for the change to take effect.";
-                    throw new ServiceBusException(message, ServiceBusFailureReason.MessageSizeExceeded);
+                    if (Log.IsDebugEnabled)
+                    {
+                        dequeueMessage.ApplicationProperties.TryGetValue(Headers.MessageId, out var messageId);
+                        Log.Debug($"Message '{messageId ?? dequeueMessage.MessageId}' is too large for the batch '{batchCount}' and will be sent individually to destination {destination}.");
+                    }
+                    messagesThatCouldntBeSent.Add(dequeueMessage);
```

Comment:
We talked with @johnsimons about separating the cases where the message is too big and the batch getting too big (and then sending the batch and starting a new one). Is that still relevant or YAGNI?

Reply:
Yes, I think we should try to batch again and again, till we cannot batch single messages, which means they are over 1MB. I hope it makes sense?

Reply:
That's what the code does. First TryAdd can only fail when the message is too large. Subsequent TryAdd attempt to add as much as possible until it returns false (either too large or batch full) then the current batch is sent, a new one is created and the message is attempted to add again. If the batch was full before but the message fits in it will be added. Otherwise it will be pulled out for individual sends and batching continues